src/server.rs

```diff
@@ -0,0 +1,230 @@
+        // TODO: decide what to do in exclusive case
+        match client_init.shared {
+            true => {}
+            false => {}
+        }
+
+        let data = self.data.lock().await;
+        let server_init = ServerInit::new(
+            data.width,
+            data.height,
+            self.config.name.clone(),
+            data.input_pixel_format.clone(),
+        );
+        info!("Tx [{:?}]: ServerInit={:#?}", addr, server_init);
+        server_init.write_to(s).await?;
+
+        Ok(())
+    }
+
+    async fn handle_conn(&self, s: &mut TcpStream, addr: SocketAddr) {
+        info!("[{:?}] new connection", addr);
+
+        if let Err(e) = self.rfb_handshake(s, addr).await {
+            error!("[{:?}] could not complete handshake: {:?}", addr, e);
+            return;
+        }
+
+        if let Err(e) = self.rfb_initialization(s, addr).await {
+            error!("[{:?}] could not complete handshake: {:?}", addr, e);
+            return;
+        }
+
+        let data = self.data.lock().await;
+        let mut output_pixel_format = data.input_pixel_format.clone();
+        drop(data);
+
+        loop {
+            let req = ClientMessage::read_from(s).await;
+
+            match req {
+                Ok(client_msg) => match client_msg {
+                    SetPixelFormat(pf) => {
+                        debug!("Rx [{:?}]: SetPixelFormat={:#?}", addr, pf);
+
+                        // TODO: invalid pixel formats?
+                        output_pixel_format = pf;
+                    }
+                    SetEncodings(e) => {
+                        debug!("Rx [{:?}]: SetEncodings={:?}", addr, e);
+                    }
+                    FramebufferUpdateRequest(f) => {
+                        debug!("Rx [{:?}]: FramebufferUpdateRequest={:?}", addr, f);
+
+                        let mut fbu = self.server.get_framebuffer_update().await;
+                        let data = self.data.lock().await;
+
+                        // We only need to change pixel formats if the client requested a different
+                        // one than what's specified in the input.
+                        //
+                        // For now, we only support transformations between 4-byte RGB formats, so
+                        // if the requested format isn't one of those, we'll just leave the pixels
+                        // as is.
+                        if data.input_pixel_format != output_pixel_format
+                            && data.input_pixel_format.is_rgb_888()
+                            && output_pixel_format.is_rgb_888()
+                        {
+                            debug!(
+                                "transforming: input={:#?}, output={:#?}",
+                                data.input_pixel_format, output_pixel_format
+                            );
+                            fbu = fbu.transform(&data.input_pixel_format, &output_pixel_format);
+                        }
```

Comment:
Add debug line for when `input != output` but one isn't rgb888?

Reply:
done